### DISCUSSION TITLE
Option to disable anon key access to database

### DISCUSSION THREAD

**Discussion by maxb94** [supabase/supabase#4547] [Questions]
I'm creating and interacting with my Supabase database using Prisma via a node backend, and use the supabase-js client on the frontend solely for auth.

I've realised that the `supabaseKey` needed for creating the client on the frontend also grants users access to the database by default. 

Is there a way to blanket disable this access, instead of going to each table in the dashboard and setting up a custom policy? Doing the latter is risky since certain Prisma commands can wipe policies on the database and I don't want to go into the Supabase dashboard after each deployment to ensure the policies are still there.

**Comment by c-p-b** [comment]
I'm rather surprised that this is open by default. It seems quite bad from a security perspective. For someone who doesn't really know what they are doing, it seems quite easy under the default settings to let anonymous users probe your API. At the very least, IMO, RLS should be enabled on each table by default. But even this is not enough - RLS enablement doesn't cover functions/RPC's. Also postgrest is configured to automatically show what you have permissions to do on OpenAPI by default - which allows any anonymous user to probe database  structure if they have permissions on it simply by going to the root and calling it. So even if you turned on RLS for your tables any anonymous user could see what functions they have privileges to execute if they have your API endpoint.   

So initially when I was trying to do this I thought this would be enough (subbing in postgres for your database name, of course):

```sql
REVOKE ALL PRIVILEGES ON DATABASE "postgres" FROM "anon";
```

That will prevent anon from getting in and querying anything. But they can still see the underlying structure of anything they have permissions on via the OpenAPI spec that is made available by default. In other words, they can't connect to the database, but they can see what they have permissions on in the database if they could. Still quite bad, IMO. So now we have to revoke all privileges on everything in the public schema and make sure nothing gets added back in when new objects are created by altering the defaults:

```sql

REVOKE ALL PRIVILEGES ON DATABASE "postgres" FROM "anon";
REVOKE ALL PRIVILEGES ON SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON SCHEMA "storage" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "storage" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "storage" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "storage" FROM "anon";
```

However, even THAT is not enough. If you do that and go check the OpenAPI specification, anonymous users can still see any functions that you have in the database, even though we revoked all privileges from them. This is because public grants execute on functions by default ( and the code that checks whether to display it in OpenAPI looks at execute:

https://hackage.haskell.org/package/postgrest-8.0.0/docs/src/PostgREST.DbStructure.html#accessibleProcs

So finally we want to revoke the ability for the public schema to execute functions. THEN all of that function info disappears from the OpenAPI spec

```sql
REVOKE ALL ON ALL ROUTINES IN SCHEMA public FROM PUBLIC;
```

Sure, I could just do the first thing and turn off database connections and also turn off openAPI for unauthenticated users, but perhaps I want to not have to mess with Postgrest config to do that.

Maybe I'm missing something here or perhaps @steve-chavez can weigh in but that seems to me, apart from deleting the role entirely (which may have other ramifications for supabase?) that revoking the privileges permanently is a fix but not really the best or most elegant fix.

**Reply by steve-chavez** [reply]
> it seems quite easy under the default settings to let anonymous users probe your API.

Note that for Supabase, Kong also requires key auth so even anonymous users need a valid JWT, the public internet cannot probe your API.

Btw, I've also recommended revoking default privileges for anon in https://github.com/supabase/supabase/discussions/1493#discussioncomment-724842 for users that want a closed by default(also PostgREST's default) db.

I must say this part is not settled yet for Supabase, so any feedback is welcome.

**Reply by c-p-b** [reply]
Thanks Steve. I actually was testing with my own ingress (Ambassador edge stack) for which I hadn't turned on key auth yet which is how I noticed this. I started with a fresh db to test again, and for reference, your revoke statements in your link worked just fine, but I still had to execute this statement to get the stored procedures/functions to disappear:

```sql
REVOKE ALL ON ALL ROUTINES IN SCHEMA public FROM PUBLIC;
```

Cheers!

**Reply by Christopher-Hayes** [reply]
Thank you for this. So, I guess for anyone who is only using Supabase edge functions (and is okay with them being enumerated) the following config seems to make sense as another layer of security on top of RLS:
```sql
REVOKE ALL PRIVILEGES ON DATABASE "postgres" FROM "anon";
REVOKE ALL PRIVILEGES ON SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON SCHEMA "storage" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "storage" FROM "anon";
-- REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "storage" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "storage" FROM "anon";
```

**Comment by gabrielsestrem** [comment]
I did all the revokes command but I still can't prevent the anon apikey from getting the entire database schema from PostgREST API https://projectid.supabase.co/rest/v1/?apikey=XXXX

Any ideas? @steve-chavez , @cpdeethree

**Reply by steve-chavez** [reply]
Try doing

```sql
ALTER ROLE authenticator SET pgrst.openapi_mode TO 'follow-privileges';

NOTIFY pgrst;
```

https://postgrest.org/en/stable/configuration.html#openapi-mode

**Reply by gabrielsestrem** [reply]
@steve-chavez This command does not work. 

But even if it works, the authenticator role is shared between the `anon` and `authenticated` users.

I want to revoke everything from the `anon` user only. I don't want to change anything to the  `authenticated` user.

**Reply by grieblius** [reply]
Hi,

I revoked from some tables and functions for `anon`  and `PUBLIC` role and it worked for me. Just need to reload PostgREST configuration. Check this answer: https://github.com/supabase/supabase/discussions/9464#discussioncomment-3893177

BTW for me it worked only in Supabase Cloud - on Local Supabase environment seems it doesn't use "in-database" configuration.

**Reply by jonathanrose9** [reply]
Hi, 

@steve-chavez In the SQL-Editor of supabase I tried to run:
```sql
ALTER ROLE authenticator SET pgrst.openapi_mode TO 'follow-privileges';
```

I got the following error message as a result:
 
`Failed to run sql query: "authenticator" is a reserved role, only superusers can modify it.`

Any ideas why it doesn't work? Thanks in advance.

**Reply by steve-chavez** [reply]
@jonathanrose9 This is why: https://github.com/supabase/supabase/discussions/9314#discussioncomment-4469017. A fix is getting into production soon.

**Reply by masda70** [reply]
`ALTER ROLE authenticator SET pgrst.openapi_mode TO 'follow-privileges';` didn't work for me, both on a local Supabase instance and Supabase Cloud, as suggested. I created a private schema e.g., `admin` and only granted access to it to the service_role, but I am unable to limit access to the OpenAPI schema spec for the anon/authenticator/authenticated roles.

`disabled` perfectly works for blocking access to the entire OpenAPI description, which is what I will use for now:

`ALTER ROLE authenticator SET pgrst.openapi_mode TO 'disabled';`

**Reply by steve-chavez** [reply]
> ALTER ROLE authenticator SET pgrst.openapi_mode TO 'follow-privileges'; didn't work for me

You need to do `NOTIFY pgrst;` afterwards.

**Reply by masda70** [reply]
I did run `NOTIFY pgrst;`, sorry I wasn't clear. The `disabled` mode actually only works with it.

**Reply by masda70** [reply]
After a bit more investigation, I managed to make the `follow-privileges` setting have some effect. I allows me to hide some or all of the RPC functions for newly created private schema `create schema admin` from the OpenAPI description. However, what I did also hid the descriptions of these functions for all roles, including the service_role :

`REVOKE ALL ON all routines in schema admin FROM  public;`

The effect of running above query can be seen by running this query:

`select * from information_schema.role_routine_grants WHERE routine_schema = 'admin';`

The REVOKE query simply removes the line that corresponds to the grantee `PUBLIC` for the functions in the `admin` schema, for which the `is_grantable` column is by default properly set to `NO`.

Regarding table visibility, I can make them visible to all roles in the OpenAPI description by doing `grant all on all tables in schema admin to anon;`, but granting table access to `service_role` only has no effect on the visibility, whether I check the OpenAPI response using either the anon or the service_role key.

I am not sure how to interpret these results. Whether this is an issue/limitation in Postgrest, or a miss configuration by Supabase. I am leaning towards an issue with Postgrest because the behavior I observe seems quite inconsistent with I would expect from changing some parameters in the config.

**Reply by steve-chavez** [reply]
> but granting table access to service_role only has no effect on the visibility, whether I check the OpenAPI response using either the anon or the service_role key.

I don't understand this. Do you mean that revoking privileges on `service_role` still shows access to tables in the openapi output with `follow-privileges`?

> I am not sure how to interpret these results. Whether this is an issue/limitation in Postgrest, or a miss configuration by Supabase. I am leaning towards an issue with Postgrest because the behavior I observe seems quite inconsistent with I would expect from changing some parameters in the config.

postgREST doesn't do anything on roles permissions, it's stateless. If you suspect a bug in postgREST, I recommend running it on a clean local postgres database (you can follow a tutorial [here](https://docs.postgrest.org/en/v12/tutorials/tut0.html)) and report an issue.

Supabase does change the default whitelist pg behavior for tables/views access on the API roles, there's a way to to revert that back with `ALTER DEFAULT PRIVILEGES` as mentioned on https://github.com/orgs/supabase/discussions/4547#discussioncomment-9064981

**Comment by gpkc** [comment]
It is actually strange that the anon key is needed in a client for logging in. I imagine this is mainly because of some required database writes (like `last_sign_in_at`), but still theoretically one should be able to obtain at least a token without the database key.

**Comment by deepak-coding-art** [comment]
Can we disable RPC on a PostgreSQL function. For do not allow users to call a function that do some critical database calculation

**Reply by steve-chavez** [reply]
You can REVOKE EXECUTE on the function or put it into a `private` schema, so it can't be reached out through the API.

**Reply by mlynch** [reply]
If you put a function in a private schema, will `rpc` fail to work? Seems like it automatically prepends `public.` to the function name when calling it. I think it's a valid usecase to use `rpc` but call functions in other schemas

**Reply by steve-chavez** [reply]
The PostgREST API exposes a schema table/views/functions, this can be configured with `db-schemas`. On Supabase this happens to be configured to `public`.

> If you put a function in a private schema, will rpc fail to work?

It simply won't find the function because PostgREST has no knowledge of it. If the schema is named `private` then you'll likely don't want it exposed to the API.

> I think it's a valid usecase to use rpc but call functions in other schemas

You can add other schemas to `db-schemas`.

**Comment by felixhagspiel** [comment]
Are there any plans to disable the public access without the hustle of disabling access for specific roles & RLS? 
I understand that supabase is build heavily around row level security and postgres functions, which works for smaller projects / MVPs, but if you want to use it as a "regular" cloud database (like the ones from heroku, digital ocean, render.com etc... with a private connection from the backend only) it seems that I have to do a lot of adjustments & workarounds. I really like the ease of setting up supabase, the integrated authentication system & the dashboard, but those things keep me from using it in bigger projects.

**Reply by kasvith** [reply]
i also need this

**Reply by mlynch** [reply]
I would also say that, after using supabase now for a while, I find I rarely use the client-side database features and would like to disable it completely to avoid any security holes. There are a lot of little things you need to think about like a user changing a column they have access to through RLS to support insert/update, but that you would be very surprised if they just changed, and RLS doesn't really have "column" level security. I'm finding I prefer just accessing my database through traditional server-side APIs

**Reply by felixhagspiel** [reply]
@steve-chavez can you share some insights, or do you know if that is (or would be) up for discussion?

**Reply by felixhagspiel** [reply]
As I doubt that this will be possible in the near future, I decided to switch to [nhost.io ](https://nhost.io/) which uses hasura under the hood.

**Reply by bombillazo** [reply]
Sad that the Supabase defaults are a hassle to revoke, especially when all we need from Supabase client is the auth. I don't want to expose/allow requests to my schema/functions just because the client exists in the frontend for auth purposes.

We, too, abandoned the Supabase client usage except for auth, and I don't want to keep exposing data and schema unnecessarily.

**Comment by bombillazo** [comment]
For anyone in the future, this is what we had to execute to get the following:

- Stop exposing the Rest API schema via the `/rest/v1/` endpoint to `anon` and `authenticated` roles
- Deny access to the `public` and `storage` schemas from the Supabase client
- Keep `service-role` functional (so far)

Caveats:
- Using the supabase client will still return errors exposing some details about the schema...

```sql

alter default privileges in schema public revoke all on functions from anon;
alter default privileges in schema public revoke all on functions from authenticated;
alter default privileges in schema public revoke all on sequences from anon;
alter default privileges in schema public revoke all on sequences from authenticated;
alter default privileges in schema public revoke all on tables from anon;
alter default privileges in schema public revoke all on tables from authenticated;
alter default privileges in schema storage revoke all on functions from anon;
alter default privileges in schema storage revoke all on functions from authenticated;
alter default privileges in schema storage revoke all on sequences from anon;
alter default privileges in schema storage revoke all on sequences from authenticated;
alter default privileges in schema storage revoke all on tables from anon;
alter default privileges in schema storage revoke all on tables from authenticated;

REVOKE ALL ON ALL ROUTINES IN SCHEMA public FROM PUBLIC;
REVOKE ALL ON ALL ROUTINES IN SCHEMA storage FROM PUBLIC;
GRANT ALL ON SCHEMA public TO "service_role";
GRANT ALL ON ALL ROUTINES IN SCHEMA storage TO "service_role";

REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "public" FROM "authenticated";
REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "storage" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "storage" FROM "authenticated";
REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "public" FROM "authenticated";
REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "storage" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "storage" FROM "authenticated";
REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "public" FROM "authenticated";
REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "storage" FROM "anon";
REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "storage" FROM "authenticated";
REVOKE ALL PRIVILEGES ON DATABASE "postgres" FROM "anon";
REVOKE ALL PRIVILEGES ON DATABASE "postgres" FROM "authenticated";
REVOKE ALL PRIVILEGES ON SCHEMA "public" FROM "anon";
REVOKE ALL PRIVILEGES ON SCHEMA "public" FROM "authenticated";
REVOKE ALL PRIVILEGES ON SCHEMA "storage" FROM "anon";
REVOKE ALL PRIVILEGES ON SCHEMA "storage" FROM "authenticated";
revoke execute on all functions in schema public from anon;
revoke execute on all functions in schema public from authenticated;
revoke execute on all functions in schema storage from anon;
revoke execute on all functions in schema storage from authenticated;
revoke select on all tables in schema public from anon;
revoke select on all tables in schema public from authenticated;
revoke select on all tables in schema storage from anon;
revoke select on all tables in schema storage from authenticated;
revoke usage on all sequences in schema public from anon;
revoke usage on all sequences in schema public from authenticated;
revoke usage on all sequences in schema storage from anon;
revoke usage on all sequences in schema storage from authenticated;
revoke usage on schema public from anon;
revoke usage on schema public from authenticated;
revoke usage on schema storage from anon;
revoke usage on schema storage from authenticated;


ALTER ROLE anon SET pgrst.openapi_mode TO 'disabled';
ALTER ROLE authenticated SET pgrst.openapi_mode TO 'disabled';
NOTIFY pgrst, 'reload config'
```

I hope it helps, and I hope this doesn't break anything related to SB now and in the long run.

**Reply by robinbdru** [reply]
Thanks @bombillazo, it works perfectly.
To anybody, use the website https://supabase-schema.vercel.app/ to check your config.

**Comment by saltcod** [comment]
We've recently made this easier to do — full doc here describing some approaches.
https://supabase.com/docs/guides/database/hardening-data-api

**Reply by eqoram** [reply]
What if you only want to deny access for the anon role? Maybe it makes sense to add the adjusted SQL commands from @bombillazo in the docs as well.

> For anyone in the future, this is what we had to execute to get the following:
> 
> * Stop exposing the Rest API schema via the `/rest/v1/` endpoint to `anon` and `authenticated` roles
> * Deny access to the `public` and `storage` schemas from the Supabase client
> * Keep `service-role` functional (so far)
> 
> Caveats:
> 
> * Using the supabase client will still return errors exposing some details about the schema...
> 
> ```sql
> alter default privileges in schema public revoke all on functions from anon;
> alter default privileges in schema public revoke all on functions from authenticated;
> alter default privileges in schema public revoke all on sequences from anon;
> alter default privileges in schema public revoke all on sequences from authenticated;
> alter default privileges in schema public revoke all on tables from anon;
> alter default privileges in schema public revoke all on tables from authenticated;
> alter default privileges in schema storage revoke all on functions from anon;
> alter default privileges in schema storage revoke all on functions from authenticated;
> alter default privileges in schema storage revoke all on sequences from anon;
> alter default privileges in schema storage revoke all on sequences from authenticated;
> alter default privileges in schema storage revoke all on tables from anon;
> alter default privileges in schema storage revoke all on tables from authenticated;
> 
> REVOKE ALL ON ALL ROUTINES IN SCHEMA public FROM PUBLIC;
> REVOKE ALL ON ALL ROUTINES IN SCHEMA storage FROM PUBLIC;
> GRANT ALL ON SCHEMA public TO "service_role";
> GRANT ALL ON ALL ROUTINES IN SCHEMA storage TO "service_role";
> 
> REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "public" FROM "anon";
> REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "public" FROM "authenticated";
> REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "storage" FROM "anon";
> REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA "storage" FROM "authenticated";
> REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "public" FROM "anon";
> REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "public" FROM "authenticated";
> REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "storage" FROM "anon";
> REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA "storage" FROM "authenticated";
> REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "public" FROM "anon";
> REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "public" FROM "authenticated";
> REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "storage" FROM "anon";
> REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA "storage" FROM "authenticated";
> REVOKE ALL PRIVILEGES ON DATABASE "postgres" FROM "anon";
> REVOKE ALL PRIVILEGES ON DATABASE "postgres" FROM "authenticated";
> REVOKE ALL PRIVILEGES ON SCHEMA "public" FROM "anon";
> REVOKE ALL PRIVILEGES ON SCHEMA "public" FROM "authenticated";
> REVOKE ALL PRIVILEGES ON SCHEMA "storage" FROM "anon";
> REVOKE ALL PRIVILEGES ON SCHEMA "storage" FROM "authenticated";
> revoke execute on all functions in schema public from anon;
> revoke execute on all functions in schema public from authenticated;
> revoke execute on all functions in schema storage from anon;
> revoke execute on all functions in schema storage from authenticated;
> revoke select on all tables in schema public from anon;
> revoke select on all tables in schema public from authenticated;
> revoke select on all tables in schema storage from anon;
> revoke select on all tables in schema storage from authenticated;
> revoke usage on all sequences in schema public from anon;
> revoke usage on all sequences in schema public from authenticated;
> revoke usage on all sequences in schema storage from anon;
> revoke usage on all sequences in schema storage from authenticated;
> revoke usage on schema public from anon;
> revoke usage on schema public from authenticated;
> revoke usage on schema storage from anon;
> revoke usage on schema storage from authenticated;
> 
> 
> ALTER ROLE anon SET pgrst.openapi_mode TO 'disabled';
> ALTER ROLE authenticated SET pgrst.openapi_mode TO 'disabled';
> NOTIFY pgrst, 'reload config'
> ```
> 
> I hope it helps, and I hope this doesn't break anything related to SB now and in the long run.

**Reply by elyobo** [reply]
@saltcod Have you done something else to the permissions there? We previously revoked access to `anon` on `public` (still works) and `storage` (now generates a lot of `WARNING (01006): no privileges could be revoked for "objects"` style warnings when doing `supabase db reset` and tests, which assert that access has been revoked, fail because no permission denied errors are thrown.

We are on `/supabase/postgres:15.6.1.127` in prod but were looking at `15.8.1.085` and this has changed between the two; is it something to do with https://github.com/orgs/supabase/discussions/34270?

**Comment by Albert-Gao** [comment]
This is the non-over-doing version which does not cause any warning.

```sql
-- Default privileges for future objects
-- Revoke all privileges from the public role to secure the database by default.
REVOKE ALL ON SCHEMA public FROM public;

-- Explicitly revoke all other permissions on all existing objects in the public schema
-- for the anon and authenticated roles. This enforces the "zero client DB access" model.
REVOKE ALL ON ALL TABLES IN SCHEMA public FROM anon, authenticated;
REVOKE ALL ON ALL FUNCTIONS IN SCHEMA public FROM anon, authenticated;
REVOKE ALL ON ALL SEQUENCES IN SCHEMA public FROM anon, authenticated;

-- Set default privileges for future objects created by the postgres role (or other admin roles).
-- to ensure they are not accessible by anon or authenticated roles by default.
ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA public
REVOKE ALL ON TABLES FROM anon, authenticated;

ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA public
REVOKE ALL ON FUNCTIONS FROM anon, authenticated;

ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA public
REVOKE ALL ON SEQUENCES FROM anon, authenticated;

-- Disable PostgREST's OpenAPI endpoint for anon and authenticated roles.
ALTER ROLE anon SET pgrst.openapi_mode TO 'disabled';
ALTER ROLE authenticated SET pgrst.openapi_mode TO 'disabled';

NOTIFY pgrst, 'reload config';
```
